## RoBERTa Model

##### roBERTa model implemented from huggingface transformers library.
> https://huggingface.co/hamzab/roberta-fake-news-classification


The model is a roberta-base fine-tuned on Bisaillon's fake-and-real-news-dataset. Initial Implemetation date: 13/07/2023.

> https://www.kaggle.com/datasets/clmentbisaillon/fake-and-real-news-dataset  

It supposedly a 100% accuracy on that dataset. The model takes a news article and predicts if it is true or fake. This model can be used in real time by inputting a headline and copy of the news article (any given length). The model can directly be implemented from the huggingface text classification model library.

In [2]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification

tokenizer = AutoTokenizer.from_pretrained("hamzab/roberta-fake-news-classification")

model = AutoModelForSequenceClassification.from_pretrained("hamzab/roberta-fake-news-classification")

#### Initial test of the model
I took a random news story from searching COVID-19 online and copied it in.

In [13]:
import torch
def predict_fake(title,text):
    input_str = "<title>" + title + "<content>" +  text + "<end>"
    input_ids = tokenizer.encode_plus(input_str, max_length=512, padding="max_length", truncation=True, return_tensors="pt")
    device =  'cuda' if torch.cuda.is_available() else 'cpu'
    model.to(device)
    with torch.no_grad():
        output = model(input_ids["input_ids"].to(device), attention_mask=input_ids["attention_mask"].to(device))
    return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))
    
print(predict_fake("Covid tests for China travellers to England ending","Travellers flying into England from mainland China will no longer have to provide proof of a negative pre-departure test from next month. The change will come into effect on 5 April - exactly three months after the measures started. Ministers brought in controls after a spike in cases following Beijing's relaxation of its zero-Covid policy. Their removal comes after greater transparency from China, the government said.In a statement, the Department of Health and Social Care (DHSC) said that there has been increased information on testing, vaccination and genomic sequencing results on China's domestic disease levels.The data indicates that Covid variants seen in China continue to be the same as those already circulating in the UK, it added.The Chinese Centre for Disease Control and Prevention also reported that all regions had passed their infection peak, the statement reported.The DHSC also announced that the UK Health Security Agency's (UKHSA) voluntary on-arrival testing programme of travellers from China to Heathrow airport has come to an end."))

# https://www.bbc.co.uk/news/business-64993197 

{'Fake': 5.5498923757113516e-05, 'Real': 0.9999444484710693}


C:\Users\luoco\AppData\Local\Temp\ipykernel_24912\2965564230.py:9: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))


##### I took an article directly from the internet as quoted above from the website, we can see that the model shows  0.9999444484710693 'Real' value. 

## Project research
For my project research, if humans are better than AI at detecting real or fake news. I will used sets of headlines that were created by LLMs. The ones chosen are guarunteed to be fake. The model is pre-trained to tale title and test however I will continue to only work with titles for the purpose of my project. Implemeted 17/08/2023.

The outputs used are:
(1) ChatGPT3.5 Output 9
(2) BARD Outputs 8.1 & 9 (20 total)
(3) HuggingChat Output 4
Please refer to Output notebooks as necessary.

The outputs used are ones which I believe are evidently fake or alternatively, where the AI has stated in the output that the headlines are false, just examples, or non genuine. So, I am looking for the model to output high scoring 'Fake' for all of these if the LLMs are assumed to be correct.

### (1) ChatGPT3.5 Output 9

In [18]:
import torch
def predict_fake(title,text):
    input_str = "<title>" + title + "<content>" +  text + "<end>"
    input_ids = tokenizer.encode_plus(input_str, max_length=512, padding="max_length", truncation=True, return_tensors="pt")
    device =  'cuda' if torch.cuda.is_available() else 'cpu'
    model.to(device)
    with torch.no_grad():
        output = model(input_ids["input_ids"].to(device), attention_mask=input_ids["attention_mask"].to(device))
    return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))
    
#print(predict_fake("",""))
print(predict_fake("Government Unveils Ambitious Plan to Modernize Public Transport Infrastructure",""))     #1
print(predict_fake("Landmark Education Reform Bill Passes Amidst Debate Over Curriculum Changes",""))        #2
print(predict_fake("National Health Service Announces Innovative Telemedicine Initiative",""))               #3
print(predict_fake("Local Community Rallies Together to Restore Historic Landmark",""))                      #4
print(predict_fake("Exclusive Interview: Leading Scientist Discusses Breakthrough in Climate Research",""))  #5
print(predict_fake("UK Economy Shows Signs of Recovery as Unemployment Rates Drop",""))                      #6
print(predict_fake("New Study Reveals Alarming Rise in Childhood Obesity Rates",""))                         #7
print(predict_fake("Royal Family Attends Charity Event, Highlights Philanthropic Efforts",""))               #8
print(predict_fake("Brexit Trade Talks Reach Critical Phase as Negotiators Seek Common Ground",""))          #9
print(predict_fake("Artificial Intelligence Adoption Grows, Raising Questions About Automation Impact",""))  #10
print(predict_fake("Nation Celebrates Cultural Diversity with Vibrant Annual Festival",""))                  #11
print(predict_fake("National Security Alert: Government Increases Counterterrorism Measures",""))            #12
print(predict_fake("Investigation Uncovers Scandal Involving Prominent Political Figure",""))                #13
print(predict_fake("Wildlife Conservation Efforts Pay Off as Endangered Species Thrive",""))                 #14
print(predict_fake("Youth-Led Climate Protest Draws Thousands, Demanding Government Action",""))             #15
print(predict_fake("Tech Startups Flourish as Entrepreneurial Spirit Gains Momentum",""))                    #16
print(predict_fake("Housing Crisis Prompts Calls for Affordable Housing Solutions",""))                      #17
print(predict_fake("Transportation Disruption: Commuters Brace for Rail Network Overhaul",""))               #18
print(predict_fake("Entertainment Industry Embraces Virtual Platforms Amidst Pandemic Challenges",""))       #19
print(predict_fake("Innovative British Invention Revolutionizes Renewable Energy Storage",""))               #20

C:\Users\luoco\AppData\Local\Temp\ipykernel_24912\1414402708.py:9: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))


{'Fake': 0.9999790191650391, 'Real': 2.09736535907723e-05}
{'Fake': 0.999980092048645, 'Real': 1.9965144019806758e-05}
{'Fake': 0.99997878074646, 'Real': 2.122170917573385e-05}
{'Fake': 0.9999792575836182, 'Real': 2.0707988369395025e-05}
{'Fake': 0.9999791383743286, 'Real': 2.082955870719161e-05}
{'Fake': 0.9999786615371704, 'Real': 2.1290852600941435e-05}
{'Fake': 0.9999785423278809, 'Real': 2.1485398974618874e-05}
{'Fake': 0.9999792575836182, 'Real': 2.073943142022472e-05}
{'Fake': 0.9999791383743286, 'Real': 2.0917725123581477e-05}
{'Fake': 0.9999799728393555, 'Real': 2.006770773732569e-05}
{'Fake': 0.9999790191650391, 'Real': 2.1001875211368315e-05}
{'Fake': 0.9999783039093018, 'Real': 2.1651387214660645e-05}
{'Fake': 0.999980092048645, 'Real': 1.993968544411473e-05}
{'Fake': 0.9999796152114868, 'Real': 2.0342647985671647e-05}
{'Fake': 0.999980092048645, 'Real': 1.9918892576242797e-05}
{'Fake': 0.9999798536300659, 'Real': 2.0198671336402185e-05}
{'Fake': 0.9999796152114868, 'Real':

#### All scores above are 0.9999 or higher for 'Fake'. This group I was certain that they were fake due to the lake of specifics on people and places.
# ______________________________________

### (2) BARD Outputs 8.1 & 9 (20 total)

In [20]:
import torch
def predict_fake(title,text):
    input_str = "<title>" + title + "<content>" +  text + "<end>"
    input_ids = tokenizer.encode_plus(input_str, max_length=512, padding="max_length", truncation=True, return_tensors="pt")
    device =  'cuda' if torch.cuda.is_available() else 'cpu'
    model.to(device)
    with torch.no_grad():
        output = model(input_ids["input_ids"].to(device), attention_mask=input_ids["attention_mask"].to(device))
    return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))
    
#print(predict_fake("",""))
print(predict_fake("Queen Elizabeth II to step down from throne after 70 years",""))                    #1
print(predict_fake("UK government to introduce mandatory vaccination for all adults",""))               #2
print(predict_fake("New study finds that climate change is causing more extreme weather events",""))    #3
print(predict_fake("Trump announces run for president in 2024",""))                                     #4
print(predict_fake("North Korea fires ballistic missile into the Sea of Japan",""))                     #5
print(predict_fake("China overtakes US as world's largest economy",""))                                 #6
print(predict_fake("EU agrees to ban all Russian oil imports",""))                                      #7
print(predict_fake("Huge asteroid to pass close to Earth next week",""))                                #8
print(predict_fake("Scientists discover new planet that could support life",""))                        #9
print(predict_fake("Major breakthrough in cancer research",""))                                         #10
print(predict_fake("Boris Johnson to resign after partygate scandal",""))                               #11
print(predict_fake("Keir Starmer to become next Prime Minister",""))                                    #12
print(predict_fake("Labour Party to win general election with landslide victory",""))                   #13
print(predict_fake("SNP to win Scottish independence referendum",""))                                   #14
print(predict_fake("UK to leave European Union without a deal",""))                                     #15
print(predict_fake("Second COVID-19 pandemic hits UK",""))                                              #16
print(predict_fake("Huge earthquake hits London",""))                                                   #17
print(predict_fake("Met Office issues 'red alert' for extreme heat",""))                                #18
print(predict_fake("UK to experience worst drought in 100 years",""))                                   #19
print(predict_fake("Wildfires destroy large swathes of countryside",""))                                #20

C:\Users\luoco\AppData\Local\Temp\ipykernel_24912\1637073960.py:9: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))


{'Fake': 0.00045229707029648125, 'Real': 0.9995476603507996}
{'Fake': 0.0002969525521621108, 'Real': 0.9997029900550842}
{'Fake': 0.9522378444671631, 'Real': 0.047762103378772736}
{'Fake': 0.0006945973145775497, 'Real': 0.9993053674697876}
{'Fake': 0.00012889583013020456, 'Real': 0.9998711347579956}
{'Fake': 0.00011262487532803789, 'Real': 0.9998873472213745}
{'Fake': 0.0002191067032981664, 'Real': 0.9997809529304504}
{'Fake': 0.957062304019928, 'Real': 0.04293769598007202}
{'Fake': 0.04002648591995239, 'Real': 0.9599735736846924}
{'Fake': 0.9988986253738403, 'Real': 0.0011013949988409877}
{'Fake': 0.00024746471899561584, 'Real': 0.9997525811195374}
{'Fake': 0.9995028972625732, 'Real': 0.0004971075686626136}
{'Fake': 0.00011431254824856296, 'Real': 0.9998856782913208}
{'Fake': 0.0013205165741965175, 'Real': 0.9986794590950012}
{'Fake': 0.0002755498280748725, 'Real': 0.9997244477272034}
{'Fake': 0.999737560749054, 'Real': 0.00026239134604111314}
{'Fake': 0.9370129108428955, 'Real': 0.06

##### Accordinging to this roBERTa model, BARD gave me true and fake content. The output from these included BARD stating 'These headlines are all plausible and could be easily mistaken for real news if they were published by a UK news agency. However, they are all false.' We cannot trust generative language models either without legitimate sources. 

##### This highlights the problem of where and who we believe. I have more trust in this roBERTa model than the generative AI as some of which state they can output things that are factually incorrect.
# ______________________________________



### (3) HuggingChat Output 4
"Hong Kong police arrest dozens in latest anti-government protests"
"Mali's Keita names new government after resignation of prime minister"
"UN Security Council extends Somalia peacekeeping mission"
"Czech PM Babis survives no-confidence vote, keeps grip on power"
"Myanmar's Suu Kyi faces pressure over Rohingya genocide allegations at ICJ"

In [22]:
import torch
def predict_fake(title,text):
    input_str = "<title>" + title + "<content>" +  text + "<end>"
    input_ids = tokenizer.encode_plus(input_str, max_length=512, padding="max_length", truncation=True, return_tensors="pt")
    device =  'cuda' if torch.cuda.is_available() else 'cpu'
    model.to(device)
    with torch.no_grad():
        output = model(input_ids["input_ids"].to(device), attention_mask=input_ids["attention_mask"].to(device))
    return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))
    
#print(predict_fake("",""))
print(predict_fake("US and China edge closer to trade deal, but tensions remain",""))                            #1
print(predict_fake("NATO chief warns of Russian threat to European security",""))                                #2
print(predict_fake("Iran nuclear deal on life support as France, Germany threaten to pull out",""))              #3
print(predict_fake("North Korea fires ballistic missile, sparking alarm in Seoul and Tokyo",""))                 #4
print(predict_fake("UK PM Johnson denies cover-up over Parliament shutdown, defends Brexit strategy",""))        #5
print(predict_fake("Pakistan PM Khan calls for dialogue with India after Kashmir clashes",""))                   #6
print(predict_fake("Putin's party wins majority in Russian parliamentary elections",""))                         #7
print(predict_fake("US Defense Secretary Esper visits Kabul, promises continued support for Afghan forces",""))  #8
print(predict_fake("Sudan's Bashir sentenced to 20 years in prison for corruption",""))                          #9
print(predict_fake("Protests erupt in Egypt after government proposes constitutional changes",""))               #10
print(predict_fake("Netanyahu's Likud Party wins Israeli election, setting up right-wing coalition",""))         #11
print(predict_fake("Thousands march in London to demand second Brexit referendum",""))                           #12
print(predict_fake("US and South Korea begin joint military exercises, angering Pyongyang",""))                  #13
print(predict_fake("Erdogan's AKP loses Istanbul mayoralty in blow to Turkish president",""))                    #14
print(predict_fake("Hong Kong police arrest dozens in latest anti-government protests",""))                      #15
print(predict_fake("Indonesia's Widodo inaugurated for second term, vows to tackle radicalism",""))              #16
print(predict_fake("Mali's Keita names new government after resignation of prime minister",""))                  #17
print(predict_fake("UN Security Council extends Somalia peacekeeping mission",""))                               #18
print(predict_fake("Czech PM Babis survives no-confidence vote, keeps grip on power",""))                        #19
print(predict_fake("Myanmar's Suu Kyi faces pressure over Rohingya genocide allegations at ICJ",""))             #20

C:\Users\luoco\AppData\Local\Temp\ipykernel_24912\2098638969.py:9: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  return dict(zip(["Fake","Real"], [x.item() for x in list(torch.nn.Softmax()(output.logits)[0])] ))


{'Fake': 7.238820398924872e-05, 'Real': 0.9999276399612427}
{'Fake': 0.00011133169027743861, 'Real': 0.9998886585235596}
{'Fake': 3.419608037802391e-05, 'Real': 0.9999657869338989}
{'Fake': 3.326481964904815e-05, 'Real': 0.9999667406082153}
{'Fake': 6.742769619449973e-05, 'Real': 0.9999325275421143}
{'Fake': 8.306757808895782e-05, 'Real': 0.9999169111251831}
{'Fake': 3.4630684240255505e-05, 'Real': 0.9999653100967407}
{'Fake': 0.00013278414553496987, 'Real': 0.9998672008514404}
{'Fake': 0.22115154564380646, 'Real': 0.77884840965271}
{'Fake': 6.767686136299744e-05, 'Real': 0.9999322891235352}
{'Fake': 8.927072485676035e-05, 'Real': 0.9999107122421265}
{'Fake': 3.624751843744889e-05, 'Real': 0.9999637603759766}
{'Fake': 6.514920096378773e-05, 'Real': 0.9999347925186157}
{'Fake': 5.127354233991355e-05, 'Real': 0.9999487400054932}
{'Fake': 3.325790385133587e-05, 'Real': 0.9999667406082153}
{'Fake': 0.00018118921434506774, 'Real': 0.9998188614845276}
{'Fake': 3.879232463077642e-05, 'Real': 

#### This output on HugginChat also stated to me 'Please note that these are just examples and not real news headlines.' at the end of the output post the headlines. 

I researched the anomaly, headline 'Sudan's Bashir sentenced to 20 years in prison for corruption', the statement overall is false, the number of years is incorrect (2 years not 20 years). Which explains why the model has a lower 'Real' accuracy. 

#### I will continue this research futher by running the outputs through an open-sourced BERT model via huggingface Transformers. This is bert-base-uncased run on colab. I can only assume what the roBERTa model has predicted is true based on the supposedly '100%' accuracy. I will use this as validation for the base-uncased model. 